src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,309 @@
+                .ToArray();
+
+
+            _compilation = CSharpCompilation.Create("CodeValidation")
+                            .AddSyntaxTrees(_syntaxTree)
+                            .AddReferences(coreLib)
+                            .AddReferences(fxLibs)
+                            .AddReferences(await NuGetReferenceProvider.GetReferences());
+
+            _compilation = _compilation.WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            _semanticModel = _compilation.GetSemanticModel(_syntaxTree, false);
+            var semanticDiagnostics = _semanticModel.GetDiagnostics();
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult, 
+                                                !ErrorExists(validationMessages), 
+                                                validationMessages);
+        }
+
+        public async Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var root = (CompilationUnitSyntax)await _syntaxTree.GetRootAsync();
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)root.Members[0];
+            var classDeclaration = namespaceDeclaration.Members
+                .OfType<ClassDeclarationSyntax>()
+                .First();
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType($"{namespaceDeclaration.Name}.{classDeclaration.Identifier.Text}");
+            var metadata = ExtractMetadata(newType);
+
+            return metadata;
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                if (algoProperty.PropertyType.IsEnum)
+                    parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                metadata.Parameters.Add(parameter);
+            }
+
+            var functionProperties = algoType
+                .GetRuntimeFields()
+                .Where(x => x.FieldType.BaseType == typeof(AbstractFunction));
+
+            foreach (var functionProperty in functionProperties)
+            {
+                var function = ToAlgoMetadataFunction(functionProperty);
+
+                var functionParamBasePropertyField = functionProperty
+                    .FieldType
+                    .GetFields()
+                    .FirstOrDefault(x => x.FieldType.BaseType == typeof(FunctionParamsBase));
```

Comment:
As said in the previous comment, perhaps using properties would be better, and also this wouldn't work if we had a middle class like. The instantiation validation should also be performed here.
```c#
public class MyParams : MyFunctionParamsBase
{}

public abstract class MyFunctionParamsBase : FunctionParamsBase
{}
```

Also, `GetFields()` includes static fields in the search as well